### PR TITLE
[SourceKit] Add range for attributes in a structure (SR-5587)

### PR DESCRIPTION
**Pull request by marcelofabri**
Adds `key.offset` and `key.length` for attributes in `key.attributes`. The SourceKit part of this PR is working already.

~However, it looks like the parser currently doesn't provide the range for most attributes (all `SimpleDeclAttr`, `ObjCAttr` and probably others).~

~For example, for this program:~

```swift
@available(swift 3.0.2)
@objc(Bar) @IBDesignable final class Foo {
  
}
```

~I get this structure:~

```
{
  key.offset: 0,
  key.length: 72,
  key.diagnostic_stage: source.diagnostic.stage.swift.parse,
  key.substructure: [
    {
      key.kind: source.lang.swift.decl.class,
      key.accessibility: source.lang.swift.accessibility.internal,
      key.name: "Foo",
      key.offset: 55,
      key.length: 16,
      key.runtime_name: "Bar",
      key.nameoffset: 61,
      key.namelength: 3,
      key.bodyoffset: 66,
      key.bodylength: 4,
      key.attributes: [
        {
          key.offset: 49,
          key.length: 0,
          key.attribute: source.decl.attribute.final
        },
        {
          key.offset: 35,
          key.length: 1,
          key.attribute: source.decl.attribute.ibdesignable
        },
        {
          key.offset: 24,
          key.length: 1,
          key.attribute: source.decl.attribute.objc.name
        },
        {
          key.offset: 0,
          key.length: 22,
          key.attribute: source.decl.attribute.available
        }
      ]
    }
  ]
}
```

~Note that only the only length that is correct is the one from `source.decl.attribute.available`.~

~I've looked into changing the parser, but that's a bigger change and I want some input if that'd be desirable in the first place. Some advice on how to do that would be great as well. I think the place that I would need to change is `bool Parser::parseNewDeclAttribute(DeclAttributes &Attributes, SourceLoc AtLoc, DeclAttrKind DK)` in `ParseDecl.cpp`.~

Resolves [SR-5587](https://bugs.swift.org/browse/SR-5587).

// cc @nkcsgexi for advice

### CODE REVIEW

**Comment by jrose-apple** [comment]
See the documentation for SourceRange: https://github.com/apple/swift/blob/master/include/swift/Basic/SourceLoc.h#L81-L84

**Comment by marcelofabri** [comment]
@jrose-apple Thanks for pointing me to the right direction! This is what I get now:

```
{
  key.offset: 0,
  key.length: 72,
  key.diagnostic_stage: source.diagnostic.stage.swift.parse,
  key.substructure: [
    {
      key.kind: source.lang.swift.decl.class,
      key.accessibility: source.lang.swift.accessibility.internal,
      key.name: "Foo",
      key.offset: 55,
      key.length: 16,
      key.runtime_name: "Bar",
      key.nameoffset: 61,
      key.namelength: 3,
      key.bodyoffset: 66,
      key.bodylength: 4,
      key.attributes: [
        {
          key.offset: 49,
          key.length: 5,
          key.attribute: source.decl.attribute.final
        },
        {
          key.offset: 35,
          key.length: 13,
          key.attribute: source.decl.attribute.ibdesignable
        },
        {
          key.offset: 24,
          key.length: 5,
          key.attribute: source.decl.attribute.objc.name
        },
        {
          key.offset: 0,
          key.length: 23,
          key.attribute: source.decl.attribute.available
        }
      ]
    }
  ]
}
```

As you can see from the structure, I'm still having issues with `ObjCAttr` though: it uses `baseRange` as its `Range`. Is this something that could/should be changed?

File: test/SourceKit/DocumentStructure/structure.swift.response [hunk 1]
```diff
@@ -985,6 +999,8 @@
       ],
       key.attributes: [
         {
+          key.offset: 1474,
+          key.length: 5,
```
Comment (marcelofabri) [inline test/SourceKit/DocumentStructure/structure.swift.response]: this should include the name

File: test/SourceKit/InterfaceGen/gen_clang_module.swift.response [hunk 2]
```diff
@@ -5975,6 +5977,8 @@ public class FooOverlayClassDerived : Foo.FooOverlayClassBase {
         key.namelength: 10,
         key.attributes: [
           {
+            key.offset: 6061,
+            key.length: 15,
```
Comment (marcelofabri) [inline test/SourceKit/InterfaceGen/gen_clang_module.swift.response]: no idea why this is 15 and not 4

**Comment by jrose-apple** [comment]
Hm, that just looks like we're passing the wrong SourceRange through in `ObjCAttr::createNullary` and `ObjCAttr::createSelector`, in Attr.cpp. (They should both use the RParenLoc as the range's end location.) Feel free to fix these and add tests for them!

File: test/SourceKit/InterfaceGen/gen_clang_module.swift.response (hunk 2 above)
Comment (marcelofabri) [reply test/SourceKit/InterfaceGen/gen_clang_module.swift.response]: So, I think this is actually a `unowned(unsafe)` and according to `Attr.def`, `key.attribute: source.decl.attribute.weak` is used for `unowned` too. So this is probably the expected behavior.

**Comment by marcelofabri** [comment]
@jrose-apple Should I add tests specifically for that? Or is the one that was updated in `SourceKit/DocumentStructure` enough?

**Comment by jrose-apple** [comment]
I think the update is fine. I'm not actually a SourceKit person, though!

**Review by nkcsgexi — APPROVED** [review]

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by marcelofabri** [comment]
I was getting this compile error on Linux:

```
/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift/tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp:1104:15: error: no matching member function for call to 'push_back'
        Attrs.push_back({AttrUID.getValue(), AttrOffset, AttrEnd - AttrOffset});
        ~~~~~~^~~~~~~~~
/usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/stl_vector.h:913:7: note: candidate function not viable: cannot convert initializer list argument to 'const value_type' (aka 'const std::tuple<SourceKit::UIdent, unsigned int, unsigned int>')
      push_back(const value_type& __x)
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/stl_vector.h:931:7: note: candidate function not viable: cannot convert initializer list argument to 'value_type' (aka 'std::tuple<SourceKit::UIdent, unsigned int, unsigned int>')
      push_back(value_type&& __x)
      ^
```

I've pushed aa41e7aa3bc44e7ccad2de874d696cfa2008728c which may fix it, but I'm kind of clueless here.

**Comment by marcelofabri** [comment]
Can someone please trigger @swift-ci?

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by marcelofabri** [comment]
Is there something blocking this from being merged? 😅

**Comment by nkcsgexi** [comment]
Merged, @marcelofabri thank you for pinging about this PR!

**Comment by marcelofabri** [comment]
Thanks for merging this @nkcsgexi! 💯

**Pull request merged**